### PR TITLE
[Fix #17] Don't add non-title entries to imenu

### LINKED ISSUE #17: Imenu entries for things that are not titles
I find the following block causes a corresponding imenu entry:
```
[%unbreakable,source,rust]
----
```

And also latexmath blocks, there are entries for [latexmath... as well as \end{equation}:
```
[latexmath#p-eqn,reftext="equation ({counter:eqn})"]
++++
\begin{equation}
p = 3 (m-1) m + 1
\end{equation}
++++
```
I suspect both are getting triggered by the two-line title detection, because if I replace "++" from adoc-two-line-title-del with "__", the latexmath entries stop showing up. 

Also a [NOTE] block will trigger an entry for the [NOTE] and end of the note block:

```
[NOTE]
====
Blah blah
This line will be an entry
====
```

### PR DESCRIPTION
**Pull request by ambihelical**
Fixes #17 by checking that 2-line titles have the same number of
under characters as the title text and doesn't look like the start
of an attribute block.

Unfortunately the length test can fail for multibyte text, and I don't
know a good solution to this in elisp.

Also check when (adoc-title-descriptor) returns nil, which it will at
times when the regexes match but more stringent tests fail, a nil list
will be inserted into the imenu list, which it doesn't like much, so
this condition is checked for as well.

Note the regex itself was not modified because it is used sometimes to
find titles with the wrong underline length, as noted in the comment
before these changes.

### CODE REVIEW

**Comment by ambihelical** [comment]
Apparently (length) will return the number of characters, not bytes, so the caveat in the commit message isn't true.  I tested some multibyte titles, they worked as would be desired.

**Comment by bbatsov** [comment]
It'd be nice if we documented somewhere the indented behavior, otherwise it's kind of hard to argue what it should be. This will also need a changelog entry.

File: adoc-mode.el [hunk 1]
```diff
@@ -2402,9 +2402,11 @@ Returns nil if there was no xref found."
                       (> saved-point (match-end 0)))))
       id)))
 
-(defun adoc-title-descriptor()
+(defun adoc-title-descriptor (&optional strict-match )
   "Returns title descriptor of title point is in.
 
+When STRICT-MATCH is t, 2 line title length must be within 2 chars of undertext
```
Comment (bbatsov) [inline adoc-mode.el:2408]: I have to admit I don't understand what this means. :D I'm guessing that undertext is whatever is under the title. Btw, aren't the title separators different from those for other blocks?

Comment (bbatsov) [reply adoc-mode.el:2408]: (I hate 2-line titles, as they require a lot of typing, and never use them myself)

**Comment by bbatsov** [comment]
Ideally we should also add a couple of unit tests for this.

File: adoc-mode.el (hunk 1 above)
Comment (ambihelical) [reply adoc-mode.el:2408]: I will try to improve the description, how does "title underlining" sound.  Honestly didn't realize they existed until looking at this package.  Adoc-mode has support for fixing the number of repeated characters, so we need to have the non-strict mode for that functionality.  

++++ and ---- are two block starts that can be confused for title underlining.  It only works in practice because 4 character titles are not usual.  

Thinking about at this now, maybe just checking for > 4 repeated characters would be sufficient for differentiating blocks from section titles.  I'll take a look at this when I fix the description. 

BTW, it would be nice if the other failing tests were fixed so it's easier to see when changes break something.   Maybe disable them until they can be fixed?

Comment (bbatsov) [reply adoc-mode.el:2408]: > BTW, it would be nice if the other failing tests were fixed so it's easier to see when changes break something. Maybe disable them until they can be fixed?

Agreed. I've opened an issue about fixing/disabling them a while ago, but I've been pretty short on time lately.

**Comment by bbatsov** [comment]
I've commented out the broken unit tests. The CI is still broken, though, due to some byte-compilation issues that we'll need to address.
